mlir/lib/Quantum/Transforms/MergeRotationsPatterns.cpp

```diff
@@ -42,59 +48,290 @@ SmallVector<mlir::Value> convertOpParamsToValues(CustomOp &op, mlir::PatternRewr
+                loc, rewriter.create<arith::AddFOp>(loc, phi1, omega1), halfConst);
+            auto alpha2 = rewriter.create<arith::MulFOp>(
+                loc, rewriter.create<arith::AddFOp>(loc, phi2, omega2), halfConst);
+            auto beta1 = rewriter.create<arith::MulFOp>(
+                loc, rewriter.create<arith::SubFOp>(loc, phi1, omega1), halfConst);
+            auto beta2 = rewriter.create<arith::MulFOp>(
+                loc, rewriter.create<arith::SubFOp>(loc, phi2, omega2), halfConst);
+
+            // c1 = cos(θ1/2), c2 = cos(θ2/2)
+            // s1 = sin(θ1/2), s2 = sin(θ2/2)
+            auto theta1Half = rewriter.create<arith::MulFOp>(loc, theta1, halfConst);
+            auto c1 = rewriter.create<math::CosOp>(loc, theta1Half);
+            auto s1 = rewriter.create<math::SinOp>(loc, theta1Half);
+            auto theta2Half = rewriter.create<arith::MulFOp>(loc, theta2, halfConst);
+            auto c2 = rewriter.create<math::CosOp>(loc, theta2Half);
+            auto s2 = rewriter.create<math::SinOp>(loc, theta2Half);
+
+            // cF = sqrt(c1^2 * c2^2 +
+            //           s1^2 * s2^2 -
+            //           2 * c1 * c2 * s1 * s2 * cos(ω1 + ϕ2))
+            auto c1TimesC2 = rewriter.create<arith::MulFOp>(loc, c1, c2);
+            auto s1TimesS2 = rewriter.create<arith::MulFOp>(loc, s1, s2);
+            auto firstAddend =
+                rewriter.create<arith::MulFOp>(loc, rewriter.create<arith::MulFOp>(loc, c1, c1),
+                                               rewriter.create<arith::MulFOp>(loc, c2, c2));
+            auto secondAddend =
+                rewriter.create<arith::MulFOp>(loc, rewriter.create<arith::MulFOp>(loc, s1, s1),
+                                               rewriter.create<arith::MulFOp>(loc, s2, s2));
+            auto thirdAddend = rewriter.create<arith::NegFOp>(
+                loc, rewriter.create<arith::MulFOp>(
+                         loc, twoConst,
+                         rewriter.create<arith::MulFOp>(
+                             loc, c1TimesC2,
+                             rewriter.create<arith::MulFOp>(
+                                 loc, s1TimesS2,
+                                 rewriter.create<math::CosOp>(
+                                     loc, rewriter.create<arith::AddFOp>(loc, omega1, phi2))))));
+            auto cF = rewriter.create<math::SqrtOp>(
+                loc, rewriter.create<arith::AddFOp>(
+                         loc, firstAddend,
+                         rewriter.create<arith::AddFOp>(loc, secondAddend, thirdAddend)));
+
+            // TODO: can we check these problematic scenarios for differentiability by code?
+            // Problematic scenarios for differentiability:
+            //
+            // 1. if (is_close_to(cF, 0)) { /* sqrt not differentiable at 0 */ return failure(); }
+            // 2. if (is_close_to(cF, 1)) { /* acos not differentiable at 1 */ return failure(); }
```

Comment:
Hmm great spot!

In general, not all programs will be differentiated though, so if we do add verification we should check it based on context. Having said that, I think it's okay to leave this as a todo for the purpose of this pr.